Add option to end the stream when to option is reached

khorwood opened the pull request:
Adds an option to stop processing the stream when the number of lines
indicated by options.to is reached. This allows for only the requested
number of lines to be parsed, preventing parsing errors from later
lines.

Depending on the expectations this may fix #132

wdavidw commented:
Thanks, this subject has been raised in the past and never yet coved. Do we really need a new option for this functionality? To me, it looks more like a bug than a missing feature. There two purpose to end processing when `to` is achieved:

* prevent error in later CSV code, as you mentionned
* speed up processing, potentially drastically if a lot of records remains to be parse

Also, could you please add a test which illustrate the first point. In current source code, the test shall fail since error will be present after to, and the test shall pass with your enhancements.

What's your opinion?

khorwood commented:
Sounds good. Would you prefer that I remove the option and make this the normal behaviour when `to` is defined?
Once we're in agreement I'll update the code and add the needed tests.

wdavidw commented:
> Sounds good. Would you prefer that I remove the option and make this the normal behaviour when to is defined?

This is my opinion. I ask you yours to make sure my arguments are valid, my understanding correct and that I'm not missing something. You can go ahead on my side.

khorwood commented:
I've removed the `endStreamWithTo` option as this feels more like a bug. Added a test where parsing the last line would result in an error, and have adjusted the parser to ignore additional calls once `to` has been reached.

khorwood commented on lib/es5/index.js:
```diff
@@ -390,6 +390,11 @@ Parser.prototype.__push = function (line) {
+  if (this.count >= this.options.to) {
+    this.isEnded = true;
+    return this.push(null);
+  }
```
Moved this block to after the push/emit to ensure the record being processed is returned.

wdavidw commented:
What is the reason of moving down those four line after emitting records ?

khorwood commented:
Moving that block allows the current parsed record to be emitted, before checking if `count >= to`, otherwise I found that the tests would lose the last result.

wdavidw commented:
Strange, the currently shouldn't pass then. Are you sure that you swith the condition `>=` back to `>` when moving the code above ?

khorwood commented:
I believe the tests are accurate.

In the original position, line 3 would get parsed when `to: 2` so a parsing error could occur if line 3 was malformed.
In the new position, after parsing and emitting line 2 it will check if `count >= 2` and then emit null to end the stream, set the `isEnded` which will prevent parsing of line 3+.

wdavidw commented:
OK, thank you for your feedback

wdavidw commented:
A patch version has been published.
